In [1]:
import numpy as np
import pandas as pd

## Dataset: A subset of all the purchases of beverages made in Thailand in a given time period

# In this notebook, we have two main purposes
### 1. <font color='red'> Clean </font> our dataset.
### 2. <font color='red'> Explore </font> and understand the dataset, for further analysis and visualisations later.

In [2]:
data = pd.read_excel("EXAMPLE_BEVERAGE_DATA_THAILAND.xlsx")

In [3]:
data.shape

(398912, 35)

In [4]:
data.head()

seqno  HOUSEID             date1             date2      Date3   weekday  \
0      1        1  1 w/e 2016/01/03  4 w/e 2016/01/03 2016-01-03  Saturday   
1      2        1  1 w/e 2016/01/03  4 w/e 2016/01/03 2016-01-03  Saturday   
2      3        1  1 w/e 2016/01/03  4 w/e 2016/01/03 2016-01-03  Thursday   
3      4        1  1 w/e 2016/01/03  4 w/e 2016/01/03 2016-01-03  Thursday   
4      5        1  1 w/e 2016/01/24  4 w/e 2016/01/31 2016-01-24  Thursday   

                                REFID         trade       channel STORENAME  \
0  REFID04203050235002000000100000000  Modern Trade  Hypermarkets     Makro   
1  REFID04203420323002000000100000000  Modern Trade  Hypermarkets     Makro   
2  REFID05102010120003000100000000000  Modern Trade  Hypermarkets     Big C   
3  REFID05102050125003000200000000000  Modern Trade  Hypermarkets     Big C   
4  REFID04202810100002000000700000000  Modern Trade  Hypermarkets     Big C   

   ...                                           SKU_DESC  CW_1Y  CW_2Y  \
0  ...  RTD TEA YEN YEN BY ICHITAN 316.6ML X6 COOL HER...    970    662   
1  ...  RTD TEA YEN JAB JAI 400ML X24 JUBLIANG HERBAL ...    970    662   
2  ...           CSD PEPSI 1000ML X12 COLA PLASTIC BOTTLE    970    662   
3  ...         CSD FANTA 1300ML STRAWBERRY PLASTIC BOTTLE    970    662   
4  ...  RTD TEA LIPTON 1000ML LEMON ICE TEA BOTTLE PLA...    970    662   

        DEMOG_AREA  DEMOG_HHINCOME   DEMOG_HHSIZE        DEMOG_LIFESTAGE  \
0  Greater Bangkok    Urban >40000  3-4 Member HH  Mixed Adults Families   
1  Greater Bangkok    Urban >40000  3-4 Member HH  Mixed Adults Families   
2  Greater Bangkok    Urban >40000  3-4 Member HH  Mixed Adults Families   
3  Greater Bangkok    Urban >40000  3-4 Member HH  Mixed Adults Families   
4  Greater Bangkok    Urban >40000  3-4 Member HH  Mixed Adults Families   

     DEMOG_HWAGE     DEMOG_REGION DEMOG_WORKINGSTATUS  
0  HWife age 50+  Greater Bangkok             Working  
1  HWife age 50+  Greater Bangkok             Working  
2  HWife age 50+  Greater Bangkok             Working  
3  HWife age 50+  Greater Bangkok             Working  
4  HWife age 50+  Greater Bangkok             Working  

[5 rows x 35 columns]

Drop **seqno** as it is just index of the rows in the dataframe. 

Drop **REFID** as it is just the serial number. 

Drop **date1** and **date2** as well (told by Professor) and just use **Date3**. 

Drop **CW_1Y** and **CW_2Y** as well (told by Professor).

In [5]:
data = data.drop(labels = ['seqno', 'date1', 'date2', 'REFID', 'CW_1Y', 'CW_2Y'], axis = 1)

Now, let us look at the type of data we have.

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398912 entries, 0 to 398911
Data columns (total 29 columns):
HOUSEID                398912 non-null int64
Date3                  398912 non-null datetime64[ns]
weekday                398912 non-null object
trade                  398912 non-null object
channel                398912 non-null object
STORENAME              398912 non-null object
packs                  398912 non-null float64
unitsperpack           398912 non-null float64
volume                 398912 non-null float64
spend                  398912 non-null float64
Priceperlitre          398912 non-null float64
Unitsizeml             398912 non-null float64
category               398912 non-null object
BRAND_E1               398640 non-null object
PACKSIZE2              398912 non-null object
packdesc               398912 non-null object
PACKSIZE_E2            398640 non-null float64
PACKAGE                398640 non-null object
FLAVOUR                258546 non-null object
F

# It seems we can roughly separate our columns into a few categories. 
1. <font color='red'> **Buyer details** </font> (i.e. **HOUSEID, Date3, weekday**) 
2. <font color='red'> **Store details** </font> (i.e. **trade, channel, STORENAME**)
3. <font color='red'> **Purchase Quantity and Spending details** </font>(i.e. **packs, unitsperpack, volume, spend, Priceperlitre, Unitsizeml, PACKSIZE_E2, PACKSIZE2**)
4. <font color='red'> **Beverage details** </font>(i.e. **category, BRAND_E1, packdesc, PACKAGE, FLAVOUR, FORMULA1, FORMULA2, SKU_DESC**)
5. <font color='red'> **Demographic of Buyer** </font> (i.e. all **DEMOG_xx**) 

Most of our data are categorical data, except for category 3.

# Check for Missing Values

Check the columns where there are missing values. Keep them in mind and deal with them later.

In [7]:
col_means = pd.DataFrame(data.isna().agg(func = np.mean), columns = ['mean'])
col_means[col_means['mean'] != 0]

mean
BRAND_E1     0.000682
PACKSIZE_E2  0.000682
PACKAGE      0.000682
FLAVOUR      0.351872
FORMULA1     0.824723
FORMULA2     0.824723

# We will not check for duplicated rows
**It is possible that a customer returns to the same store on the same day and make the exact same purchase.**

# <font color='red'> 1. Inspect the Buyer details

Let us rename all the columns to camel case for consistency.

In [8]:
data = data.rename(columns = {'HOUSEID' :'houseId', 
                              'Date3' : 'date'})

In [9]:
data[['houseId', 'date', 'weekday']].head()

houseId       date   weekday
0        1 2016-01-03  Saturday
1        1 2016-01-03  Saturday
2        1 2016-01-03  Thursday
3        1 2016-01-03  Thursday
4        1 2016-01-24  Thursday

Recall that there are no missing values for this category. Thus, we will simply look at the columns to get a general idea of the dataset.

Look at **houseId**.

In [10]:
print(len(np.unique(data['houseId'])))
print(min(np.unique(data['houseId'])))
print(max(np.unique(data['houseId'])))

5409
1
5409


There are 5409 unique customers. And since the **houseId** starts at 1 and ends at 5409, we know that each customer has an **houseId** between 1 and 5409, and all integer from 1 to 5409 represents a **houseId**.

Next, look at **date**.

In [11]:
np.unique(data['date'])[0:10]

array(['2015-11-15T00:00:00.000000000', '2015-11-22T00:00:00.000000000',
       '2015-11-29T00:00:00.000000000', '2015-12-06T00:00:00.000000000',
       '2015-12-13T00:00:00.000000000', '2015-12-20T00:00:00.000000000',
       '2015-12-27T00:00:00.000000000', '2016-01-03T00:00:00.000000000',
       '2016-01-10T00:00:00.000000000', '2016-01-17T00:00:00.000000000'],
      dtype='datetime64[ns]')

In [12]:
data['date'].dt.dayofweek.value_counts()

6    398912
Name: date, dtype: int64

It seems that all the dates are Sundays. It is likely that the purchase details are entered into the system on Sundays, and thus **date** represents the date the data is entered, but not the date the purchases are made.

Let's check that the **weekday** column is right.

In [13]:
data['weekday'].value_counts()

Sunday       60432
Monday       60169
Saturday     58277
Tuesday      56487
Wednesday    56264
Friday       53793
Thursday     53490
Name: weekday, dtype: int64

Yes, there are only the 7 days of the week in this column. We can use this column to help us correctly record the date of the purchases.

In [14]:
offset = np.zeros(len(np.array(data['weekday'])))

for i, day in enumerate(np.array(data['weekday'])):
    if (day == "Saturday"):
        offset[i] = 1
    elif (day == "Friday"):
        offset[i] = 2
    elif (day == "Thursday"):
        offset[i] = 3
    elif (day == "Wednesday"):
        offset[i] = 4
    elif (day == "Tuesday"):
        offset[i] = 5
    elif (day == "Monday"):
        offset[i] = 6
        
date = np.zeros((len(np.array(data['weekday']))), dtype='datetime64[s]')
for i in range(len(data['date'])):
    date[i] = data['date'][i] - pd.DateOffset(offset[i])

Insert this date of purchase into the dataframe, and rename the original date column.

In [15]:
data.insert(2, 'datePurchase', date)
data = data.rename(columns = {'date' : 'dateRecorded'})
data.head()

houseId dateRecorded datePurchase   weekday         trade       channel  \
0        1   2016-01-03   2016-01-02  Saturday  Modern Trade  Hypermarkets   
1        1   2016-01-03   2016-01-02  Saturday  Modern Trade  Hypermarkets   
2        1   2016-01-03   2015-12-31  Thursday  Modern Trade  Hypermarkets   
3        1   2016-01-03   2015-12-31  Thursday  Modern Trade  Hypermarkets   
4        1   2016-01-24   2016-01-21  Thursday  Modern Trade  Hypermarkets   

  STORENAME  packs  unitsperpack  volume  ...  FORMULA1  FORMULA2  \
0     Makro    1.0           6.0     1.9  ...       NaN       NaN   
1     Makro    1.0          24.0     9.6  ...       NaN       NaN   
2     Big C    1.0          12.0    12.0  ...       NaN       NaN   
3     Big C    1.0           1.0     1.3  ...       NaN       NaN   
4     Big C    1.0           1.0     1.0  ...       NaN       NaN   

                                            SKU_DESC       DEMOG_AREA  \
0  RTD TEA YEN YEN BY ICHITAN 316.6ML X6 COOL HER...  Greater Bangkok   
1  RTD TEA YEN JAB JAI 400ML X24 JUBLIANG HERBAL ...  Greater Bangkok   
2           CSD PEPSI 1000ML X12 COLA PLASTIC BOTTLE  Greater Bangkok   
3         CSD FANTA 1300ML STRAWBERRY PLASTIC BOTTLE  Greater Bangkok   
4  RTD TEA LIPTON 1000ML LEMON ICE TEA BOTTLE PLA...  Greater Bangkok   

  DEMOG_HHINCOME   DEMOG_HHSIZE        DEMOG_LIFESTAGE    DEMOG_HWAGE  \
0   Urban >40000  3-4 Member HH  Mixed Adults Families  HWife age 50+   
1   Urban >40000  3-4 Member HH  Mixed Adults Families  HWife age 50+   
2   Urban >40000  3-4 Member HH  Mixed Adults Families  HWife age 50+   
3   Urban >40000  3-4 Member HH  Mixed Adults Families  HWife age 50+   
4   Urban >40000  3-4 Member HH  Mixed Adults Families  HWife age 50+   

      DEMOG_REGION DEMOG_WORKINGSTATUS  
0  Greater Bangkok             Working  
1  Greater Bangkok             Working  
2  Greater Bangkok             Working  
3  Greater Bangkok             Working  
4  Greater Bangkok             Working  

[5 rows x 30 columns]

We are all good for the first category of columns. Now, let's inspect the store details.

# <font color='red'> 2. Inspect the Store details

Rename all the columns to camel case for consistency.

In [16]:
data = data.rename(columns = {'STORENAME' : 'storeName'})

In [17]:
data[['trade', 'channel', 'storeName']].head()

trade       channel storeName
0  Modern Trade  Hypermarkets     Makro
1  Modern Trade  Hypermarkets     Makro
2  Modern Trade  Hypermarkets     Big C
3  Modern Trade  Hypermarkets     Big C
4  Modern Trade  Hypermarkets     Big C

Let us look at the type of categories there are in **trade** and **channel**.

In [18]:
data['trade'].value_counts()

Traditional Trade    181240
Modern Trade         146075
Others                71597
Name: trade, dtype: int64

In [19]:
data['channel'].value_counts()

Provision St      181240
Convenience St     91659
Direct Sales       47843
Hypermarkets       37110
Others             23754
Supermarkets       17306
Name: channel, dtype: int64

**trade** is likely to be a more general category as compared to **channel**, in describing a store. 

Let us look at the type of **channel** there are in each of the 3 categories of **trade**.

In [20]:
data[data['trade'] == 'Traditional Trade']['channel'].value_counts()

Provision St    181240
Name: channel, dtype: int64

In [21]:
data[data['trade'] == 'Modern Trade']['channel'].value_counts()

Convenience St    91659
Hypermarkets      37110
Supermarkets      17306
Name: channel, dtype: int64

In [22]:
data[data['trade'] == 'Others']['channel'].value_counts()

Direct Sales    47843
Others          23754
Name: channel, dtype: int64

The results are logical and already cleaned. We will retain the **trade** column for possible usage in the data visualisation later on, even though the column does not provide much additional information.

Inspect **storeName**.

In [23]:
data['storeName'].value_counts()

Provision store          181240
7-Eleven                  63806
Drinking water agency     45841
Tesco Lotus               19286
Tesco lotus express       12687
                          ...  
TPI mart                      1
Central                       1
First super                   1
Thathong                      1
Tiger mart (Esso)             1
Name: storeName, Length: 261, dtype: int64

At the top, we can see that there are many beverages purchases made at provision stores. We also have big chains of convenience stores/hypermarkets, having many beverages purchases.

Let us check if all provision stores do not have a **storeName** recorded.

In [24]:
data[data['channel'] == 'Provision St']['storeName'].value_counts()

Provision store    181240
Name: storeName, dtype: int64

Indeed, all the provision stores (~50% rows of the dataset) do not have unique **storeName**s.

# <font color='red'> 3. Inspect the Purchase quantities and spending details

In [25]:
data[['packs', 'unitsperpack', 'volume', 'spend', 'Priceperlitre', 'Unitsizeml', 'PACKSIZE2', 'PACKSIZE_E2']].head()

packs  unitsperpack  volume  spend  Priceperlitre   Unitsizeml   PACKSIZE2  \
0    1.0           6.0     1.9   45.0      23.684211   316.666667  316.6ML X6   
1    1.0          24.0     9.6  175.0      18.229167   400.000000   400ML X24   
2    1.0          12.0    12.0  230.0      19.166667  1000.000000  1000ML X12   
3    1.0           1.0     1.3   21.0      16.153846  1300.000000      1300ML   
4    1.0           1.0     1.0   31.0      31.000000  1000.000000      1000ML   

   PACKSIZE_E2  
0       1890.0  
1       9600.0  
2      12000.0  
3       1250.0  
4       1000.0

We see that **PACKSIZE2** and **PACKSIZE_E2** does not introduce new information, since we already have **unitsperpack** and **volume**. Furthermore, there are missing values in **PACKSIZE_E2**. 

Thus, we will just drop these 2 columns altogether.

In [26]:
data = data.drop(labels = ['PACKSIZE2', 'PACKSIZE_E2'], axis = 1)

In [27]:
data[['packs', 'unitsperpack', 'volume', 'spend', 'Priceperlitre', 'Unitsizeml']].head()

packs  unitsperpack  volume  spend  Priceperlitre   Unitsizeml
0    1.0           6.0     1.9   45.0      23.684211   316.666667
1    1.0          24.0     9.6  175.0      18.229167   400.000000
2    1.0          12.0    12.0  230.0      19.166667  1000.000000
3    1.0           1.0     1.3   21.0      16.153846  1300.000000
4    1.0           1.0     1.0   31.0      31.000000  1000.000000

These 6 columns are clearly related, but the naming convention can be confusing. Also, the scale is different (some in millilitres and some in litres)

In [28]:
# To convert Unitsizeml to scale of Litres, and change the naming
loc = int(np.where(data.columns == 'unitsperpack')[0][0]) + 1
data.insert(loc, 'litresPerUnit', data['Unitsizeml'] / 1000)
data = data.drop(['Unitsizeml'], axis = 1)

# Rename volume to litres and unitsperpack to unitsPerPack
data = data.rename(columns = {'volume' : 'litres', 'unitsperpack' : 'unitsPerPack'})

# Move position of Priceperlitre and lowercase
loc = int(np.where(data.columns == 'litresPerUnit')[0][0]) + 1
data.insert(loc, 'pricePerLitre',  data['Priceperlitre'])
data = data.drop(['Priceperlitre'], axis = 1)

data[['packs', 'unitsPerPack', 'litresPerUnit', 'pricePerLitre', 'litres', 'spend']].head()

packs  unitsPerPack  litresPerUnit  pricePerLitre  litres  spend
0    1.0           6.0       0.316667      23.684211     1.9   45.0
1    1.0          24.0       0.400000      18.229167     9.6  175.0
2    1.0          12.0       1.000000      19.166667    12.0  230.0
3    1.0           1.0       1.300000      16.153846     1.3   21.0
4    1.0           1.0       1.000000      31.000000     1.0   31.0

Now, the columns are arranged in a proper fashion. 

To obtain **litres**, we can do **packs** * **unitsPerPack** * **litresPerUnit**. 

To obtain **spend**, we can do **packs** * **unitsPerPack** * **litresPerUnit** * **pricePerLitre**.

We keep both columns **litres** and **spend** even when they do not provide additional information, for convenience sake.

**packs** and **unitsPerPack** should be integers. Check if there are rows where the values are not integers.

In [29]:
print(np.where((data['packs'] % data['packs'].astype(int)) != 0))
print(np.where((data['unitsPerPack'] % data['unitsPerPack'].astype(int)) != 0))

(array([], dtype=int64),)
(array([210959, 212784, 314320], dtype=int64),)


Drop the 3 rows where **unitsPerPack** are not integer values, and then convert all values in both columns into integer type.

In [30]:
rows = list(np.where((data['unitsPerPack'] % data['unitsPerPack'].astype(int)) != 0)[0])
data = data.drop(rows, axis = 0)
data['packs'] = data['packs'].astype(int)
data['unitsPerPack'] = data['unitsPerPack'].astype(int)
data = data.reset_index(drop=True)

data[['packs', 'unitsPerPack', 'litresPerUnit', 'pricePerLitre', 'litres', 'spend']].head()

packs  unitsPerPack  litresPerUnit  pricePerLitre  litres  spend
0      1             6       0.316667      23.684211     1.9   45.0
1      1            24       0.400000      18.229167     9.6  175.0
2      1            12       1.000000      19.166667    12.0  230.0
3      1             1       1.300000      16.153846     1.3   21.0
4      1             1       1.000000      31.000000     1.0   31.0

# <font color='red'> 4. Inspect the Beverage details

Recall that we have many missing values in these columns too.

In [31]:
col_means = pd.DataFrame(data.isna().agg(func = np.mean), columns = ['mean'])
col_means[col_means['mean'] != 0]

mean
BRAND_E1  0.000682
PACKAGE   0.000682
FLAVOUR   0.351867
FORMULA1  0.824722
FORMULA2  0.824722

We see that a huge proportion of the data is missing for the **FLAVOUR**, **FORMULA1** and **FORMULA2** columns.

Let us look at these columns first.

In [32]:
data[['category', 'BRAND_E1', 'packdesc', 'PACKAGE', 'FLAVOUR', 'FORMULA1', 'FORMULA2', 'SKU_DESC']].head()

category            BRAND_E1           packdesc         PACKAGE  \
0  RTD TEA  YEN YEN BY ICHITAN  316.66666667ML X6  BOTTLE PLASTIC   
1  RTD TEA         YEN JAB JAI          400ML X24  BOTTLE PLASTIC   
2      CSD               PEPSI         1000ML X12  PLASTIC BOTTLE   
3      CSD               FANTA             1300ML  PLASTIC BOTTLE   
4  RTD TEA              LIPTON             1000ML  BOTTLE PLASTIC   

           FLAVOUR FORMULA1 FORMULA2  \
0    COOL HERB TEA      NaN      NaN   
1  JUBLIANG HERBAL      NaN      NaN   
2             COLA      NaN      NaN   
3       STRAWBERRY      NaN      NaN   
4    LEMON ICE TEA      NaN      NaN   

                                            SKU_DESC  
0  RTD TEA YEN YEN BY ICHITAN 316.6ML X6 COOL HER...  
1  RTD TEA YEN JAB JAI 400ML X24 JUBLIANG HERBAL ...  
2           CSD PEPSI 1000ML X12 COLA PLASTIC BOTTLE  
3         CSD FANTA 1300ML STRAWBERRY PLASTIC BOTTLE  
4  RTD TEA LIPTON 1000ML LEMON ICE TEA BOTTLE PLA...

We can immediately see that **packdesc** is a repetition of data from **litresPerUnit** and **unitsPerPack**. 

We also see that **SKU_DESC** repeats the quantity details, and also contain the details of the beverage in a "dirty" format. The details of the beverage can already be found in the category, **BRAND_E1**, **PACKAGE** and **FLAVOUR** columns.

Thus, we will discard **packdesc** and **SKU_DESC**.

In [33]:
data = data.drop(['packdesc', 'SKU_DESC'], axis = 1)
data[['category', 'BRAND_E1', 'PACKAGE', 'FLAVOUR', 'FORMULA1', 'FORMULA2']].head()

category            BRAND_E1         PACKAGE          FLAVOUR FORMULA1  \
0  RTD TEA  YEN YEN BY ICHITAN  BOTTLE PLASTIC    COOL HERB TEA      NaN   
1  RTD TEA         YEN JAB JAI  BOTTLE PLASTIC  JUBLIANG HERBAL      NaN   
2      CSD               PEPSI  PLASTIC BOTTLE             COLA      NaN   
3      CSD               FANTA  PLASTIC BOTTLE       STRAWBERRY      NaN   
4  RTD TEA              LIPTON  BOTTLE PLASTIC    LEMON ICE TEA      NaN   

  FORMULA2  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN

We will look at the type of values in the first 3 columns.

In [34]:
print(data['category'].value_counts())
print("-------")
print(data['BRAND_E1'].value_counts())
print("-------")
print(data['PACKAGE'].value_counts())

WATER      140118
CSD        137616
JUICE       70112
RTD TEA     51063
Name: category, dtype: int64
-------
OTHERS BRAND    50011
PEPSI           42646
COCA COLA       39898
FANTA           22816
OISHI           18774
                ...  
NATION              1
SOONTRA             1
FAMILY MART         1
OLE CUTE            1
D ICE               1
Name: BRAND_E1, Length: 582, dtype: int64
-------
PLASTIC BOTTLE    236224
GALLON             56399
BOTTLE PLASTIC     42169
CARTON             32931
CAN                13090
GLASS BOTTLE       12352
GABLE               2814
CUP                 2570
BOTTLE GLASS          60
BAG                   26
SACHET                 2
Name: PACKAGE, dtype: int64


From **category**, there are 4 kinds of categories of beverages. 

From **BRAND_E1**, there are 582 brands of beverages, although there is a huge number that is recorded as "OTHERS BRAND". 

From **PACKAGE**, there are 11 kinds of packaging. However, we notice that there are _PLASTIC BOTTLE_, _BOTTLE PLASTIC_, _GLASS BOTTLE_ and _BOTTLE GLASS_ as 4 of the 11 categories in this column. _PLASTIC BOTTLE_ and _BOTTLE PLASTIC_ should refer to the same kind of packaging, and _GLASS BOTTLE_ and _BOTTLE GLASS_ refers to the same kind of packaging as well. This is most likely due to mistakes made by the data collector, and thus we should rectify this.

In [35]:
packages = np.zeros(len(np.array(data['PACKAGE'])), dtype='object')
for i, package in enumerate(np.array(data['PACKAGE'])):
    if (package == "BOTTLE PLASTIC"):
        packages[i] = "PLASTIC BOTTLE"
    elif (package == "BOTTLE GLASS"):
        packages[i] = "GLASS BOTTLE"
    else:
        packages[i] = package

loc = int(np.where(data.columns == 'PACKAGE')[0][0])
data.insert(loc, 'package', packages)
data = data.drop(['PACKAGE'], axis = 1)

In [36]:
data['package'].value_counts()

PLASTIC BOTTLE    278393
GALLON             56399
CARTON             32931
CAN                13090
GLASS BOTTLE       12412
GABLE               2814
CUP                 2570
BAG                   26
SACHET                 2
Name: package, dtype: int64

The **package** column is now rectified. 

Recall that **BRAND_E1** and **package** have some missing values (0.0682% of the data). We will impute the missing values with _MISSING_.

In [37]:
from sklearn.impute import SimpleImputer

# Initialise Imputer for categorical features
imputer = SimpleImputer(strategy = 'constant', fill_value = 'MISSING')

# Performs the imputation and returns a new Numpy array
data_imputed = imputer.fit_transform(data[['BRAND_E1', 'package']])

# Enter the data back into the dataframe
data['BRAND_E1'] = data_imputed[:, 0]
data['package'] = data_imputed[:, 1]

In [38]:
col_means = pd.DataFrame(data.isna().agg(func = np.mean), columns = ['mean'])
col_means[col_means['mean'] != 0]

mean
FLAVOUR   0.351867
FORMULA1  0.824722
FORMULA2  0.824722

Now, we are done with the imputation and the last 3 columns we have yet to inspect, each have a significant number of missing values. 

Let us look at the type of values they have.

In [39]:
data['FLAVOUR'].value_counts()

COLA                                           86038
GREEN TEA HONEY LEMON                          17407
STRAWBERRY                                     11566
ORANGE                                          8897
COOL HERB TEA                                   6661
                                               ...  
25% ORANGE JUICE WITH ORANGE PULP CONLLAGEN        1
98% TOMATO JUICE                                   1
100% CRANBERRY JUICE                               1
GREEN TEA PLUM                                     1
CONCENTRATE BLUEBERRY JUICE                        1
Name: FLAVOUR, Length: 929, dtype: int64

I have an intution, that perhaps the 35% of the **FLAVOUR** column that has missing values, is because the **category** of beverage is WATER. 

Let us inspect that.

In [40]:
print(np.mean(data[data['category'] == 'WATER']['FLAVOUR'].isna()))
print(np.mean(data[data['category'] == 'CSD']['FLAVOUR'].isna()))
print(np.mean(data[data['category'] == 'JUICE']['FLAVOUR'].isna()))
print(np.mean(data[data['category'] == 'RTD TEA']['FLAVOUR'].isna()))

1.0
0.00022526450412742704
0.0027384755819260614
0.00043084033448876876


Since that is indeed the case, we will impute the missing values as _NOFLAVOUR_ if the **category** is _WATER_. Else, we will fill it as _MISSING_.

In [41]:
flavours = np.zeros(len(np.array(data['FLAVOUR'])), dtype='object')
for i in range(len(np.array(data['FLAVOUR']))):
    if data['category'][i] == 'WATER':
        flavours[i] = 'NOFLAVOUR'
    elif not isinstance(data['FLAVOUR'][i], str):
        flavours[i] = 'MISSING'
    else:
        flavours[i] = data['FLAVOUR'][i]
        
data['FLAVOUR'] = flavours

Now, look at the last two columns.

In [42]:
print(data['FORMULA1'].value_counts())
print("-------")
print(data['FORMULA2'].value_counts())

Non   100%    50106
100 %         19814
Name: FORMULA1, dtype: int64
-------
RTD        69635
NON RTD      285
Name: FORMULA2, dtype: int64


It seems like they have the same count

In [43]:
sum(data['FORMULA1'].isna() == False) == sum(data['FORMULA2'].isna() == False)

True

Let us look at where these values belong to. It is likely to belong to beverages from the _JUICE_ **category**, since the other 3 categories are all RTD already (Ready to Drink).

In [44]:
print(data['category'].value_counts())
print("----------------------------------")
print(data[data['FORMULA1'] == '100 %']['category'].value_counts())
print("----------------------------------")
print(data[data['FORMULA1'] == 'Non   100%']['category'].value_counts())
print("----------------------------------")
print(data[data['FORMULA2'] == 'RTD']['category'].value_counts())
print("----------------------------------")
print(data[data['FORMULA2'] == 'NON RTD']['category'].value_counts())

WATER      140118
CSD        137616
JUICE       70112
RTD TEA     51063
Name: category, dtype: int64
----------------------------------
JUICE    19814
Name: category, dtype: int64
----------------------------------
JUICE    50106
Name: category, dtype: int64
----------------------------------
JUICE    69635
Name: category, dtype: int64
----------------------------------
JUICE    285
Name: category, dtype: int64


Indeed, the **FORMULA1** and **FORMULA2** columns are describing beverages that are of the _JUICE_ **category**.

For **FORMULA1**, if the **category** is _JUICE_ and there are missing values, we will impute it as _MISSING_. If the **category** is not _JUICE_, we will impute the values as _NA_.

For **FORMULA2**, if the **category** is _JUICE_ and there are missing values, we will impute it as _MISSING_. If the **category** is not _JUICE_, we will impute the values as _RTD_ (because _WATER_ , _CSD_ and _RTD TEA_ are all _RTD_ ).

In [45]:
formula1 = np.zeros(len(np.array(data['FORMULA1'])), dtype='object')
formula2 = np.zeros(len(np.array(data['FORMULA2'])), dtype='object')

for i in range(len(np.array(data['FORMULA1']))):
    if data['category'][i] != 'JUICE':
        formula1[i] = 'NA'
        formula2[i] = 'RTD'
    else:
        if (not isinstance(data['FORMULA1'][i], str)):
            formula1[i] = 'MISSING'
        else:
            formula1[i] = data['FORMULA1'][i]
        
        if (not isinstance(data['FORMULA2'][i], str)):
            formula2[i] = 'MISSING'
        else:
            formula2[i] = data['FORMULA2'][i]

data['FORMULA1'] = formula1
data['FORMULA2'] = formula2

Let us check if we imputed it correctly.

In [46]:
print(data['FORMULA1'].value_counts())
print("----------------------------------")
print(data['FORMULA2'].value_counts())

NA            328797
Non   100%     50106
100 %          19814
MISSING          192
Name: FORMULA1, dtype: int64
----------------------------------
RTD        398432
NON RTD       285
MISSING       192
Name: FORMULA2, dtype: int64


There are no more missing values in the dataframe.

In [47]:
col_means = pd.DataFrame(data.isna().agg(func = np.mean), columns = ['mean'])
col_means[col_means['mean'] != 0]

Empty DataFrame
Columns: [mean]
Index: []

In [48]:
data = data.rename(columns = {'BRAND_E1' : 'brand',
                              'FLAVOUR' : 'flavour',
                              'FORMULA1' : 'concentration',
                              'FORMULA2' : 'RTD'})
data.iloc[:, 12:].head()

spend category               brand         package          flavour  \
0   45.0  RTD TEA  YEN YEN BY ICHITAN  PLASTIC BOTTLE    COOL HERB TEA   
1  175.0  RTD TEA         YEN JAB JAI  PLASTIC BOTTLE  JUBLIANG HERBAL   
2  230.0      CSD               PEPSI  PLASTIC BOTTLE             COLA   
3   21.0      CSD               FANTA  PLASTIC BOTTLE       STRAWBERRY   
4   31.0  RTD TEA              LIPTON  PLASTIC BOTTLE    LEMON ICE TEA   

  concentration  RTD       DEMOG_AREA DEMOG_HHINCOME   DEMOG_HHSIZE  \
0            NA  RTD  Greater Bangkok   Urban >40000  3-4 Member HH   
1            NA  RTD  Greater Bangkok   Urban >40000  3-4 Member HH   
2            NA  RTD  Greater Bangkok   Urban >40000  3-4 Member HH   
3            NA  RTD  Greater Bangkok   Urban >40000  3-4 Member HH   
4            NA  RTD  Greater Bangkok   Urban >40000  3-4 Member HH   

         DEMOG_LIFESTAGE    DEMOG_HWAGE     DEMOG_REGION DEMOG_WORKINGSTATUS  
0  Mixed Adults Families  HWife age 50+  Greater Bangkok             Working  
1  Mixed Adults Families  HWife age 50+  Greater Bangkok             Working  
2  Mixed Adults Families  HWife age 50+  Greater Bangkok             Working  
3  Mixed Adults Families  HWife age 50+  Greater Bangkok             Working  
4  Mixed Adults Families  HWife age 50+  Greater Bangkok             Working

# <font color='red'> 5. Inspect the Demographic of Buyers

In [49]:
data[['DEMOG_AREA', 'DEMOG_HHINCOME', 'DEMOG_HHSIZE', 'DEMOG_LIFESTAGE', 
      'DEMOG_HWAGE', 'DEMOG_REGION', 'DEMOG_WORKINGSTATUS']].head()

DEMOG_AREA DEMOG_HHINCOME   DEMOG_HHSIZE        DEMOG_LIFESTAGE  \
0  Greater Bangkok   Urban >40000  3-4 Member HH  Mixed Adults Families   
1  Greater Bangkok   Urban >40000  3-4 Member HH  Mixed Adults Families   
2  Greater Bangkok   Urban >40000  3-4 Member HH  Mixed Adults Families   
3  Greater Bangkok   Urban >40000  3-4 Member HH  Mixed Adults Families   
4  Greater Bangkok   Urban >40000  3-4 Member HH  Mixed Adults Families   

     DEMOG_HWAGE     DEMOG_REGION DEMOG_WORKINGSTATUS  
0  HWife age 50+  Greater Bangkok             Working  
1  HWife age 50+  Greater Bangkok             Working  
2  HWife age 50+  Greater Bangkok             Working  
3  HWife age 50+  Greater Bangkok             Working  
4  HWife age 50+  Greater Bangkok             Working

In [50]:
print(data['DEMOG_AREA'].value_counts())
print("----------------------------------")
print(data['DEMOG_HHINCOME'].value_counts())
print("----------------------------------")
print(data['DEMOG_HHSIZE'].value_counts())
print("----------------------------------")
print(data['DEMOG_LIFESTAGE'].value_counts())
print("----------------------------------")
print(data['DEMOG_HWAGE'].value_counts())
print("----------------------------------")
print(data['DEMOG_REGION'].value_counts())
print("----------------------------------")
print(data['DEMOG_WORKINGSTATUS'].value_counts())

Upcountry Rural    156941
Upcountry Urban    136590
Greater Bangkok    105378
Name: DEMOG_AREA, dtype: int64
----------------------------------
Urban 17500-40000    99344
Urban >40000         85361
Rural >15000         83662
Rural 6000-15000     61033
Urban < 17500        57263
Rural <6000          12246
Name: DEMOG_HHINCOME, dtype: int64
----------------------------------
3-4 Member HH    179992
2 Member HH       89400
5-6 Member HH     74632
1 Member HH       39931
7+ Member HH      14954
Name: DEMOG_HHSIZE, dtype: int64
----------------------------------
Families w Child 4-12        117725
Families w Teenager 13-19     68648
Mixed Adults Families         59557
Older Couples                 40012
Families w Baby 0-3           39009
Younger Couples               34027
Younger Singles               20184
Older Singles                 19747
Name: DEMOG_LIFESTAGE, dtype: int64
----------------------------------
HWife age 35-49    179671
HWife age 50+      167702
HWife age <35       51536

For the columns **DEMOG_HHSIZE**, **DEMOG_LIFESTAGE**, **DEMOG_HWAGE** and **DEMOG_WORKINGSTATUS**, the categories are relatively straightforward and thus we will not augment the data.

We will look at **DEMOG_AREA**, **DEMOG_HHINCOME** and **DEMOG_REGION**. It seems like there is some relationship between them.

In [51]:
print(data[data['DEMOG_AREA'] == 'Upcountry Rural']['DEMOG_HHINCOME'].value_counts())
print("----------------------------------")
print(data[data['DEMOG_AREA'] == 'Upcountry Urban']['DEMOG_HHINCOME'].value_counts())
print("----------------------------------")
print(data[data['DEMOG_AREA'] == 'Greater Bangkok']['DEMOG_HHINCOME'].value_counts())
print("----------------------------------")

Rural >15000        83662
Rural 6000-15000    61033
Rural <6000         12246
Name: DEMOG_HHINCOME, dtype: int64
----------------------------------
Urban 17500-40000    56392
Urban >40000         43163
Urban < 17500        37035
Name: DEMOG_HHINCOME, dtype: int64
----------------------------------
Urban 17500-40000    42952
Urban >40000         42198
Urban < 17500        20228
Name: DEMOG_HHINCOME, dtype: int64
----------------------------------


Yes, it seems like for the category _Upcountry Rral_ in **DEMOG_AREA**, all the household incomes (**DEMOG_HHINCOME**) are categorised under the Rural bracket. Whereas for _Upcountry Urban_ and _Greater Bangkok_, all the household incomes are categorised under the Urban bracket.

In [52]:
print(data[data['DEMOG_REGION'] == 'Greater Bangkok']['DEMOG_AREA'].value_counts())
print("----------------------------------")
print(data[data['DEMOG_REGION'] == 'North-East']['DEMOG_AREA'].value_counts())
print("----------------------------------")
print(data[data['DEMOG_REGION'] == 'Central']['DEMOG_AREA'].value_counts())
print("----------------------------------")
print(data[data['DEMOG_REGION'] == 'North']['DEMOG_AREA'].value_counts())
print("----------------------------------")
print(data[data['DEMOG_REGION'] == 'South']['DEMOG_AREA'].value_counts())

Greater Bangkok    105378
Name: DEMOG_AREA, dtype: int64
----------------------------------
Upcountry Urban    48986
Upcountry Rural    42441
Name: DEMOG_AREA, dtype: int64
----------------------------------
Upcountry Rural    41021
Upcountry Urban    32275
Name: DEMOG_AREA, dtype: int64
----------------------------------
Upcountry Rural    38266
Upcountry Urban    29143
Name: DEMOG_AREA, dtype: int64
----------------------------------
Upcountry Rural    35213
Upcountry Urban    26186
Name: DEMOG_AREA, dtype: int64


It seems like the relationship in this dataset is that Thailand is considered to have 5 regions. For each of the 4 regions other than Greater Bangkok, there is the Upcountry Rural and Upcountry Urban areas of the region. For the Greater Bangkok region, there is only considered to be one area, which is also Greater Bangkok, and is considered to be Urban.

We will augment the **DEMOG_HHINCOME** into _LOW_ , _MID_ and _HIGH_ . 
1. _LOW_ : 'Urban < 17500' and 'Rural <6000'
2. _MID_ : 'Urban 17500-40000' and 'Rural 6000-15000'
3. _HIGH_ : 'Urban >40000' and 'Rural >15000'

In [53]:
income = np.zeros(len(np.array(data['DEMOG_HHINCOME'])), dtype='object')

for i in range(len(np.array(data['DEMOG_HHINCOME']))):
    if data['DEMOG_HHINCOME'][i] in ('Urban < 17500', 'Rural <6000'):
        income[i] = 'LOW'
    elif data['DEMOG_HHINCOME'][i] in ('Urban 17500-40000', 'Rural 6000-15000'):
        income[i] = 'MID'
    elif data['DEMOG_HHINCOME'][i] in ('Urban >40000', 'Rural >15000'):
        income[i] = 'HIGH'

data['DEMOG_HHINCOME'] = income

We will also augment the **DEMOG_AREA**. 

We will convert all _Greater Bangkok_ and _Upcountry Urban_ to _Urban_ , and convert all _Upcountry Rural_ to _Rural_.


In [54]:
area = np.zeros(len(np.array(data['DEMOG_AREA'])), dtype='object')

for i in range(len(np.array(data['DEMOG_AREA']))):
    if data['DEMOG_AREA'][i] in ('Upcountry Urban', 'Greater Bangkok'):
        area[i] = 'Urban'
    elif data['DEMOG_AREA'][i] == 'Upcountry Rural':
        area[i] = 'Rural'

data['DEMOG_AREA'] = area

Rename and reorder the columns.

In [55]:
data = data.rename(columns = {'DEMOG_AREA' : 'demogArea',
                              'DEMOG_HHINCOME' : 'demogHhincome',
                              'DEMOG_HHSIZE' : 'demogHhsize',
                              'DEMOG_LIFESTAGE' : 'demogLifestage',
                              'DEMOG_HWAGE' : 'demogHwage',
                              'DEMOG_REGION' : 'demogRegion',
                              'DEMOG_WORKINGSTATUS' : 'demogWorkingStatus'})

cols = list(data.columns)
beforeDemog = cols[0 : cols.index('demogArea')]
reorgDemog = ['demogRegion', 'demogArea', 'demogWorkingStatus', 'demogHhincome', 
              'demogHhsize', 'demogLifestage', 'demogHwage']
beforeDemog.extend(reorgDemog)
data = data[beforeDemog]
data.head()

houseId dateRecorded datePurchase   weekday         trade       channel  \
0        1   2016-01-03   2016-01-02  Saturday  Modern Trade  Hypermarkets   
1        1   2016-01-03   2016-01-02  Saturday  Modern Trade  Hypermarkets   
2        1   2016-01-03   2015-12-31  Thursday  Modern Trade  Hypermarkets   
3        1   2016-01-03   2015-12-31  Thursday  Modern Trade  Hypermarkets   
4        1   2016-01-24   2016-01-21  Thursday  Modern Trade  Hypermarkets   

  storeName  packs  unitsPerPack  litresPerUnit  ...          flavour  \
0     Makro      1             6       0.316667  ...    COOL HERB TEA   
1     Makro      1            24       0.400000  ...  JUBLIANG HERBAL   
2     Big C      1            12       1.000000  ...             COLA   
3     Big C      1             1       1.300000  ...       STRAWBERRY   
4     Big C      1             1       1.000000  ...    LEMON ICE TEA   

   concentration  RTD      demogRegion demogArea demogWorkingStatus  \
0             NA  RTD  Greater Bangkok     Urban            Working   
1             NA  RTD  Greater Bangkok     Urban            Working   
2             NA  RTD  Greater Bangkok     Urban            Working   
3             NA  RTD  Greater Bangkok     Urban            Working   
4             NA  RTD  Greater Bangkok     Urban            Working   

  demogHhincome    demogHhsize         demogLifestage     demogHwage  
0          HIGH  3-4 Member HH  Mixed Adults Families  HWife age 50+  
1          HIGH  3-4 Member HH  Mixed Adults Families  HWife age 50+  
2          HIGH  3-4 Member HH  Mixed Adults Families  HWife age 50+  
3          HIGH  3-4 Member HH  Mixed Adults Families  HWife age 50+  
4          HIGH  3-4 Member HH  Mixed Adults Families  HWife age 50+  

[5 rows x 26 columns]

# We are finally done with cleaning the data!

Additionally, we will randomly add Cities to the dataset according to their **demogRegion** (source: https://en.wikipedia.org/wiki/Regions_of_Thailand). This is for visualisation purposes in Tableau. We will be grouping the cities in Tableau.

In [56]:
rand_cities = np.zeros(len(np.array(data['demogRegion'])), dtype='object')

northeast = 'Amnat Charoen, Bueng Kan, Buri Ram, Chaiyaphum, Kalasin, Khon Kaen, Loei, Maha Sarakham, Mukdahan, Nakhon Phanom, Nakhon Ratchasima, Nong Bua Lamphu, Nong Khai, Roi Et, Sakon Nakhon, Si Sa Ket, Surin, Ubon Ratchathani, Udon Thani, Yasothon'
northeast = northeast.split(", ")

north = 'Chiang Mai, Chiang Rai, Lampang, Lamphun, Mae Hong Son, Nan, Phayao, Phrae, Uttaradit, Tak, Sukhothai, Phitsanulok, Phichit, Kamphaeng Phet, Phetchabun, Nakhon Sawan, Uthai Thani'
north = north.split(", ")

central = 'Ang Thong, Phra Nakhon Si Ayutthaya, Chai Nat, Lop Buri, Nakhon Pathom, Nonthaburi, Pathum Thani, Samut Prakan, Samut Sakhon, Samut Songkhram, Saraburi, Sing Buri, Suphan Buri, Nakhon Nayok, Chachoengsao, Chanthaburi, Chon Buri, Prachin Buri, Rayong, Sa Kaeo, Trat, Kanchanaburi, Ratchaburi, Phetchaburi, Prachuap Khiri Khan'
central = central.split(", ")

south = 'Chumphon, Nakhon Si Thammarat, Narathiwat, Pattani, Phatthalung, Songkhla, Surat Thani, Yala, Krabi, Phang Nga, Phuket, Ranong, Satun, Trang'
south = south.split(", ")

In [57]:
ne_counter = 0
n_counter = 0
c_counter = 0
s_counter = 0

for i in range(len(np.array(data['demogRegion']))):
    if data['demogRegion'][i] == 'Greater Bangkok':
        rand_cities[i] = 'Bangkok'
    elif data['demogRegion'][i] == 'North-East':
        rand_cities[i] = northeast[ne_counter % len(northeast)]
        ne_counter += 1
    elif data['demogRegion'][i] == 'North':
        rand_cities[i] = north[n_counter % len(north)]
        n_counter += 1
    elif data['demogRegion'][i] == 'Central':
        rand_cities[i] = central[c_counter % len(central)]
        c_counter += 1
    elif data['demogRegion'][i] == 'South':
        rand_cities[i] = south[s_counter % len(south)]
        s_counter += 1
        
data['randomCities'] = rand_cities

## Save the dataset and move on to Tableau!

In [58]:
data.to_excel('BEVERAGES_cleaned_randCities.xlsx')